src/main/java/hudson/plugins/mercurial/MercurialSCM.java

```diff
@@ -225,7 +230,12 @@ private String getSource(EnvVars env) {
     }
 
     @Override public String getKey() {
-        return "hg " + getSource(new EnvVars()) + " " + revision;
+        String base = "hg " + getSource(new EnvVars());
+        if (revisionType == RevisionType.CHANGESET) {
+            return base;
+        } else {
+            return base + " " + revision;
```

Comment:
This is the important part. Previously `getKey` was returning a different key for each build of the branch project.